dm/dm/master/openapi_controller.go

```diff
@@ -77,85 +131,323 @@ func (s *Server) deleteSource(ctx context.Context, sourceName string, force bool
+		needUpdate = true
+	}
+	if req.RelayBinlogGtid != nil && sourceCfg.RelayBinlogGTID != *req.RelayBinlogGtid {
+		sourceCfg.RelayBinlogGTID = *req.RelayBinlogGtid
+		needUpdate = true
+	}
+	if req.RelayDir != nil && sourceCfg.RelayDir != *req.RelayDir {
+		sourceCfg.RelayDir = *req.RelayDir
+		needUpdate = true
+	}
+	if needUpdate {
+		// update current source relay config before start relay
+		if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
+			return err
+		}
+	}
+	return s.scheduler.StartRelay(sourceName, *req.WorkerNameList)
+}
+
 // nolint:unparam
-func (s *Server) listSource(ctx context.Context, req interface{}) []openapi.Source {
-	// TODO(ehco) implement filter later
-	sourceM := s.scheduler.GetSourceCfgs()
-	openapiSourceList := make([]openapi.Source, 0, len(sourceM))
-	for _, source := range sourceM {
-		openapiSourceList = append(openapiSourceList, config.SourceCfgToOpenAPISource(source))
-	}
-	return openapiSourceList
+func (s *Server) disableRelay(ctx context.Context, sourceName string, req openapi.DisableRelayRequest) error {
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	if req.WorkerNameList == nil {
+		worker := s.scheduler.GetWorkerBySource(sourceName)
+		if worker == nil {
+			return terror.ErrWorkerNoStart.Generate()
+		}
+		req.WorkerNameList = &openapi.WorkerNameList{worker.BaseInfo().Name}
+	}
+	return s.scheduler.StopRelay(sourceName, *req.WorkerNameList)
+}
+
+func (s *Server) purgeRelay(ctx context.Context, sourceName string, req openapi.PurgeRelayRequest) error {
+	purgeReq := &workerrpc.Request{
+		Type:       workerrpc.CmdPurgeRelay,
+		PurgeRelay: &pb.PurgeRelayRequest{Filename: req.RelayBinlogName},
+	}
+	if req.RelayDir != nil {
+		purgeReq.PurgeRelay.SubDir = *req.RelayDir
+	}
+	// NOTE not all worker that enabled relay is recorded in scheduler, we need refine this later
+	workers, err := s.scheduler.GetRelayWorkers(sourceName)
+	if err != nil {
+		return err
+	}
+	if len(workers) == 0 {
+		return terror.ErrOpenAPICommonError.Generatef("relay worker for source %s not found, please `enable-relay` first", sourceName)
+	}
+	for _, w := range workers {
+		resp, err := w.SendRequest(ctx, purgeReq, s.cfg.RPCTimeout)
+		if err != nil {
+			return err
+		}
+		if resp.PurgeRelay.Msg != "" {
+			return terror.ErrOpenAPICommonError.Generate(resp.PurgeRelay.Msg)
+		}
+	}
+	return nil
 }
 
-// nolint:unparam,unused
-func (s *Server) enableSource(ctx context.Context, sourceName, workerName string) error {
+func (s *Server) enableSource(ctx context.Context, sourceName string) error {
+	cfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if cfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
 	worker := s.scheduler.GetWorkerBySource(sourceName)
 	if worker == nil {
-		return terror.ErrWorkerNoStart
+		return terror.ErrWorkerNoStart.Generate()
```

Comment:
do we allow enable a source without worker bounded? 

in `disableSource` below we can disable a source without worker

Reply:
The current implementation of enable is to batch open the subtask on this source, so a worker must be available

But disable is not necessarily the case. Assuming that a source is not bouned by a worker, normally there is no bound subtask on the source